### PR TITLE
[DIP-498] Add support for manifest gate configs

### PR DESCRIPTION
**Pull request by sawidis**

### CODE REVIEW

**Comment by sawidis** [comment]
Given my time constraints, I haven't adjusted kong.extra_apis() but will do next week.

**Comment by sawidis** [comment]
(keep in mind this is my first interaction with rust and shipcat so I know that's far from ideal)

File: shipcat_cli/src/get.rs [hunk 1]
```diff
@@ -149,12 +150,24 @@ pub fn apistatus(conf: &Config, reg: &Region) -> Result<()> {
...
+            if let Some(g) = mf.gate {
+                // `manifest.verify` ensures that if there is a gate conf,
+                // `gate.public` must be equal to `publiclyAccessible`.
+                // That means that the following line does not alter the value
+                // of `params.publiclyAccessible` but will be useful during the
+                // migration of manifest configuration (ie deprecate
+                // `publiclyAccessible` in favour of `gate.public`).
+                params.publiclyAccessible = g.public;
+                params.websockets = g.websockets;
+            }
```
Comment (clux) [inline shipcat_cli/src/get.rs]: i guess this avoid any breakages to `get apistatus` - as long as you clean it up after you've revamped the consumer of that.

File: shipcat_cli/src/get.rs [hunk 2]
```diff
@@ -165,6 +178,7 @@ pub fn apistatus(conf: &Config, reg: &Region) -> Result<()> {
             hosts: api.hosts.unwrap_or("".into()),
             internal: api.internal,
             publiclyAccessible: api.publiclyAccessible,
+            websockets: false,
```
Comment (clux) [inline shipcat_cli/src/get.rs]: leave a note here about no extra_api support for gate perhaps

File: shipcat_definitions/src/structs/gate.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use std::ops::Not;
+
+/// Gate setup for a service
```
Comment (clux) [inline shipcat_definitions/src/structs/gate.rs]: /// Gate service configuration
///
/// Gate is a babylon specific, filtering entry-point for kong, as such, requires kong.
/// Configuration for `gate` is expected to be picked up outside of shipcat for services using kong.
```

File: shipcat_definitions/src/structs/gate.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+#[derive(Serialize, Deserialize, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct Gate {
+    /// let external traffic in or not
+    #[serde(default, skip_serializing_if = "Not::not")]
+    pub public: bool,
+
+    /// allow connection upgrade to websockets
```
Comment (clux) [inline shipcat_definitions/src/structs/gate.rs]: capital starting letter on comments plz

File: shipcat_definitions/src/manifest.rs [hunk 5]
```diff
@@ -783,6 +787,15 @@ impl Manifest {
             region.versioningScheme.verify(v)?;
         }
 
+        if let Some(g) = &self.gate {
+            if self.kong.is_none() {
+                bail!("Can't have a `gate` configuration without a `kong` one");
+            }
+            if g.public != self.publiclyAccessible {
+                bail!("[Migration plan] `publiclyAccessible` and `gate.public` must be equal");
+            }
```
Comment (clux) [inline shipcat_definitions/src/manifest.rs]: this is fine as long as you find a way to remove the gate param from the kong config later.

File: shipcat_cli/src/get.rs (hunk 1 above)
Comment (sawidis) [reply shipcat_cli/src/get.rs]: I'll have it cleaned up by the end of next week for sure. I tested `get apistatus` by running `make gates` in manifests and no change was rendered.

**Pull request merged**